chanfitness/chaneventstore.go

```diff
@@ -0,0 +1,381 @@
+// channel's uptime and a blocking response channel on which the result is sent.
+type uptimeRequest struct {
+	channelID    uint64
+	startTime    time.Time
+	endTime      time.Time
+	responseChan chan uptimeResponse
+}
+
+// uptimeResponse contains the response to an uptimeRequest and an error if one
+// occurred.
+type uptimeResponse struct {
+	uptime time.Duration
+	err    error
+}
+
+// NewChannelEventStore initializes an event store with the config provided.
+// Note that this function does not start the main event loop, Start() must be
+// called.
+func NewChannelEventStore(config *Config) *ChannelEventStore {
+	store := &ChannelEventStore{
+		cfg:              config,
+		channels:         make(map[uint64]*chanEventLog),
+		peers:            make(map[[33]byte]bool),
+		lifespanRequests: make(chan lifespanRequest),
+		uptimeRequests:   make(chan uptimeRequest),
+		quit:             make(chan struct{}),
+	}
+
+	return store
+}
+
+// Start adds all existing open channels to the event store and starts the main
+// loop which records channel and peer events, and serves requests for
+// information from the store.
+func (c *ChannelEventStore) Start() error {
+	// Create a subscription to channel events.
+	channelClient, err := c.cfg.ChannelEvents()
+	if err != nil {
+		return err
+	}
+
+	// Create a subscription to peer events.
+	peerClient, err := c.cfg.PeerEvents()
+	if err != nil {
+		return err
+	}
+
+	// Add the existing set of channels to the event store. This is required
+	// because channel events will not be triggered for channels that exist
+	// at startup time.
+	channels, err := c.cfg.GetOpenChannels()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Adding %v channels to event store", len(channels))
+
+	for _, ch := range channels {
+		var peerKey [33]byte
+		copy(peerKey[:], ch.IdentityPub.SerializeCompressed())
+
+		// Add existing channels to the channel store with an initial peer
+		// online or offline event.
+		c.addChannel(ch.ShortChanID().ToUint64(), peerKey)
+	}
+
+	// Start a goroutine that consumes events from all subscriptions.
+	c.wg.Add(1)
+	go c.consume(&subscriptions{
+		channelUpdates: channelClient.Updates(),
+		peerUpdates:    peerClient.Updates(),
+		cancel: func() {
+			channelClient.Cancel()
+			peerClient.Cancel()
+		},
+	})
+
+	return nil
+}
+
+// Stop terminates all goroutines started by the event store.
+func (c *ChannelEventStore) Stop() {
+	log.Info("Stopping event store")
+
+	// Stop the consume goroutine.
+	close(c.quit)
+
+	c.wg.Wait()
+}
+
+// addChannel adds a new channel to the ChannelEventStore's map of channels with
+// an initial peer online state (if the peer is online). If the channel is
+// already present in the map, the function returns early. This function should
+// be called to add existing channels on startup and when open channel events
+// are observed.
+func (c *ChannelEventStore) addChannel(channelID uint64, peer [33]byte) {
+	_, ok := c.channels[channelID]
+	if ok {
+		return
```

Comment:
Can this happen? Log something, or explain how it can happen in a comment?

Reply:
This would only happen in the case of a bug, but I thought it was worth adding the check in case. 
I'll add a log and a comment to explain? It can also be removed, since the duplicate would just be overwritten but I don't think there's much harm leaving the check in?

Reply:
Just `return` while continuing would work too doesn't sound very useful. I can imagine that this sub-system is sensitive to timing issues / race conditions. In that case, it may be nice to at least log it. Unless we are very certain this will really never happen. Don't need to log here necessarily, can also bubble up the error a bit. What works best.

Reply:
Oh, seeing now that you added the log already. Should be `Errorf` if it is a bug